inst/include/dust/rng.hpp

```diff
@@ -12,67 +12,82 @@ namespace dust {
 
 // This is just a container class for state
 template <typename T>
-class pRNG { // # nocov
+class pRNG {
 public:
-  pRNG(const size_t n, const std::vector<uint64_t>& seed) {
-    rng_state_t<T> s;
+  pRNG(const size_t n, const std::vector<uint64_t>& seed) :
+    n_(n), state_(n * rng_state_t<T>::size()) {
     auto len = rng_state_t<T>::size();
     auto n_seed = seed.size() / len;
     for (size_t i = 0; i < n; ++i) {
+      rng_state_t<T> s = state(i);
       if (i < n_seed) {
-        std::copy_n(seed.begin() + i * len, len, s.state.begin());
+        for (size_t j = 0; j < len; ++j) {
+          s[j] = seed[i * len + j];
+        }
       } else {
+        rng_state_t<T> prev = state(i - 1);
+        for (size_t j = 0; j < len; ++j) {
+          s[j] = prev[j];
+        }
         xoshiro_jump(s);
       }
-      _state.push_back(s);
     }
   }
 
   size_t size() const {
-    return _state.size();
+    return n_;
   }
 
   void jump() {
-    for (size_t i = 0; i < _state.size(); ++i) {
-      xoshiro_jump(_state[i]);
+    for (size_t i = 0; i < n_; ++i) {
+      xoshiro_jump(state(i));
     }
   }
 
   void long_jump() {
-    for (size_t i = 0; i < _state.size(); ++i) {
-      xoshiro_long_jump(_state[i]);
+    for (size_t i = 0; i < n_; ++i) {
+      xoshiro_long_jump(state(i));
     }
   }
 
-  rng_state_t<T>& state(size_t i) {
-    return _state[i];
+  // Access an individual rng_state by constructing the small struct
+  // that contains a pointer to the memory, offset as needed, and our
+  // stride.
+  rng_state_t<T> state(size_t i) {
+    return rng_state_t<T>(state_.data() + i, n_);
   }
 
+  // Possibly nicer way of doing the above
+  rng_state_t<T> operator[](size_t i) {
+    return rng_state_t<T>(state_.data() + i * n_, n_);
+  }
+
+  // We might make this optionally dump out the raw state, at least
+  // for debugging?
   std::vector<uint64_t> export_state() {
-    std::vector<uint64_t> state;
-    const size_t n = rng_state_t<T>::size();
-    state.reserve(size() * n);
-    for (size_t i = 0; i < size(); ++i) {
-      for (size_t j = 0; j < n; ++j) {
-        state.push_back(_state[i][j]);
+    const auto len = rng_state_t<T>::size();
+    std::vector<uint64_t> ret(n_ * len);
+    for (size_t i = 0; i < n_; ++i) {
+      for (size_t j = 0; j < len; ++j) {
+        ret[i * len + j] = state_[i + n_ * j];
```

Comment:
This de-interleaves? Can you note that in the comment above, I wasn't immediately sure what raw state would mean

Reply:
Added a comment